### PR TITLE
Move ForkAndExecProcess shim from coreclr to corefx

### PR DESCRIPTION
**Pull request by nguerrera**
Moved from [coreclr/src/pal/src/misc/corefx.cpp](https://github.com/dotnet/coreclr/blob/b4333eff478fd48292d350f675be2ba916ea8271/src/pal/src/misc/corefx.cpp#L206)

cc @stephentoub

### CODE REVIEW

File: src/Native/System.Native/pal_exec.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    if ((processId = fork()) == -1)
+    {
+        assert(!"fork() failed.");
+        success = false;
+        goto done;
+    }
+
+    /* From the time the child process (processId == 0) begins running from fork to when
+     * it reaches execve, the child process must not touch anything in the PAL.  Doing so
+     * is not safe. The parent process (processId >= 0) may continue to use the PAL.
```
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: This PAL comment is no longer relevant, right?

File: src/Native/System.Native/pal_exec.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+enum
+{
+    READ_END_OF_PIPE = 0,
+    WRITE_END_OF_PIPE = 1,
+};
+    
+inline static void CloseIfOpen(int fd)
+{
+    if (fd >= 0)
+        close(fd);
```
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: Maybe add a comment that we're explicitly choosing to ignore any close failures.

File: src/Native/System.Native/pal_exec.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+extern "C"
```
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: Lest anyone think it'd be a good idea to move this function back to managed code,I think a small comment here would be helpful, indicating why we can't do the fork/exec combination from managed.  Essentially, forking only clones the current thread, not all threads in the process, which makes it very unsafe given things like GC, finalization, etc.  We could, for example, end up in deadlock situations by trying to run managed code in the child process between the fork and the exec.

File: src/Native/System.Native/pal_exec.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    /* From the time the child process (processId == 0) begins running from fork to when
+     * it reaches execve, the child process must not touch anything in the PAL.  Doing so
+     * is not safe. The parent process (processId >= 0) may continue to use the PAL.
+     */
+
+    if (processId == 0) // processId == 0 if this is child process
+    {
+        // Close the parent end of any open pipes
```
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: I realize I wrote this comment in the first place :), but we should probably clarify it slightly to say that we're closing the child's copy of the parent's end of any open pipes, not that we're actually closing the parent's end.

File: src/Native/System.Native/pal_exec.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        CloseIfOpen(stdinFds[WRITE_END_OF_PIPE]);
+        CloseIfOpen(stdoutFds[READ_END_OF_PIPE]);
+        CloseIfOpen(stderrFds[READ_END_OF_PIPE]);
+
+        *stdinFd  = -1;
+        *stdoutFd = -1;
+        *stderrFd = -1;
+        *childPid = -1;
+    }
+
```
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: The intention here was that if the fork failed, errno would still be set to whatever it was from the fork call.  But I just noticed we're making intervening calls to close, and it's possible close is changing errno.  Consider capturing errno after the fork and resetting it here.

File: src/Native/System.Native/pal_exec.cpp (hunk 3 above)
Comment (stephentoub) [inline src/Native/System.Native/pal_exec.cpp]: (I realize I already had such a comment in the .h file :) Nevermind.)

**Comment by stephentoub** [comment]
A few minor comments.  Otherwise, LGTM.

File: src/Native/System.Native/pal_exec.cpp (hunk 3 above)
Comment (nguerrera) [inline src/Native/System.Native/pal_exec.cpp]: :) I've gone in the other direction and removed the pasted comments from the header for the parameters (did not mean to duplicate them) so the impl signature looks more bare and hints more at needing to look at the header for docs...

File: src/Native/System.Native/pal_exec.cpp (hunk 5 above)
Comment (nguerrera) [inline src/Native/System.Native/pal_exec.cpp]: I opted to make CloseIfOpen preserve errno which fits well with earlier suggestion to document that ignoring close errors is deliberate.

**Comment by stephentoub** [comment]
Updates LGTM.

**Pull request merged**